src/lib/protocols/present-proof/ProofService.ts

```diff
@@ -0,0 +1,914 @@
+        );
+      }
+      // If there are no restrictions we can self attest the attribute
+      else {
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        const value = credentialMatch.credentialInfo.attributes.get(requestedAttribute.name!);
+
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        requestedCredentials.selfAttestedAttributes.set(referent, value!);
+      }
+    }
+
+    for (const [referent, requestedPredicate] of proofRequest.requestedPredicates) {
+      const credentials = await this.getCredentialsForProofRequest(proofRequest, referent);
+
+      // Can't create requestedPredicates without matching credentials
+      if (credentials.length === 0) {
+        throw new Error(
+          `Could not automatically construct requested credentials for proof request '${proofRequest.name}'`
+        );
+      }
+
+      const credentialMatch = credentials[0];
+      if (requestedPredicate.restrictions) {
+        requestedCredentials.requestedPredicates.set(
+          referent,
+          new RequestedPredicate({
+            credentialId: credentialMatch.credentialInfo.referent,
+          })
+        );
+      }
+      // If there are no restrictions we can self attest the attribute
+      else {
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        const value = credentialMatch.credentialInfo.attributes.get(requestedPredicate.name!);
+
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        requestedCredentials.selfAttestedAttributes.set(referent, value!);
+      }
+    }
+
+    return requestedCredentials;
+  }
+
+  /**
+   * Verify an indy proof object. Will also verify raw values against encodings.
+   *
+   * @param proofRequest The proof request to use for proof verification
+   * @param proofJson The proof object to verify
+   * @throws {Error} If the raw values do not match the encoded values
+   * @returns Boolean whether the proof is valid
+   *
+   */
+  public async verifyProof(proofJson: IndyProof, proofRequest: ProofRequest): Promise<boolean> {
+    const proof = JsonTransformer.fromJSON(proofJson, PartialProof);
+
+    for (const [referent, attribute] of proof.requestedProof.revealedAttributes.entries()) {
+      if (!CredentialUtils.checkValidEncoding(attribute.raw, attribute.encoded)) {
+        throw new Error(
+          `The encoded value for '${referent}' is invalid. ` +
+            `Expected '${CredentialUtils.encode(attribute.raw)}'. ` +
+            `Actual '${attribute.encoded}'`
+        );
+      }
+    }
+
+    // TODO: pre verify proof json
+    // I'm not 100% sure how much indy does. Also if it checks whether the proof requests matches the proof
+    // @see https://github.com/hyperledger/aries-cloudagent-python/blob/master/aries_cloudagent/indy/sdk/verifier.py#L79-L164
+
+    const schemas = await this.buildSchemas(new Set(proof.identifiers.map(i => i.schemaId)));
```

Comment:
I would say this is about "getting" schemas rather than "building" 🤔  

To my mind, LedgerService could be extracted away from this service, and schemas passed as a method parameter. What do you think?

Reply:
Took this from .NET. Fine by me. Same as the others, will resolve in next PR. Too much changes already

Reply:
I will rename from build to get, but leaving the abstraction out for now. That would mean both ProofsModule and PresentationHandler would need to implement that method